[espmilighthub] Initial contribution (#9218)
openhab/openhab-addons: bundles/org.openhab.binding.mqtt.espmilighthub/README.md

| `fut091` | Remote is the newer 8 group model called a fut091 and your globes are cct. |
| `rgbw` | Remote is 4 channels and the globes have RGB and a fixed white. |
| `rgb` | Remote is 4 channels and the globes have full RGB with no white. |

## Discovery

First install the MQTT binding and setup a `broker` and make sure it is ONLINE, as this binding uses the MQTT binding to talk to your broker and hence that binding must be setup first.
Next make sure you have moved a control on either a physical remote, or used a virtual control inside the esp8266 control panel web page to create a MQTT message.
This binding should then detect the new device the moment the control is moved and a new entry should appear in your INBOX.

To remove a saved state from your MQTT broker that causes an entry in your INBOX you can use this command or use the ignore feature of openHAB.

```
mosquitto_pub -u username -P password -p 1883 -t 'milight/states/0x0/rgb_cct/1' -n -r
```

## Thing Configuration

| Parameter | Description | Required | Default |
|-|-|-|-|
| `defaultCommand` | When the send command button is pressed, if no selection is made from the command channel then use this. | Y | night_mode |

> | `whiteHue` | When both the `whiteHue` and `whiteSat` values are seen by the binding it will trigger the white LEDS. Set to -1 to disable, 0 for Alexa, or 36 for Google Home. | Y | 35 |

Author: This looks to be a typo in the readme, as it should be 36 for the default for what Google sends as the HUE when asked for white. Will resolve this after I confirm and fix the typo.

> | `whiteHue` | When both the `whiteHue` and `whiteSat` values are seen by the binding it will trigger the white LEDS. Set to -1 to disable, 0 for Alexa, or 36 for Google Home. | Y | 35 |

Author: FIXED: was meant to be 35 so fixed typo.

| `powerFailsToMinimum` | If lights loose power from the power switch OR a power outage, they will default to using the lowest brightness if the light was turned off before the power failure occurred. | Y | true |
| `whiteThreshold` | RGBW globes do not respond to saturation changes, so this feature allows you to specify a number that if the saturation drops below, it will trigger the white mode. -1 will disable this feature. | Y | 12 |

## Channels

| Channel | Type | Description |
|-|-|-|
| level | Dimmer | Level changes the brightness of the globe. |
| colourTemperature | Dimmer | Change from cool to warm white with this control. |
| colour | Color | Allows you to change the colour, brightness and saturation of the globe. |
| discoMode | String | Switch to a Disco mode directly from a drop down list. |
| bulbMode | String (read only) | Displays the mode the bulb is currently in so that rules can determine if the globe is white, a color, disco modes or night mode are selected. |

> | command | String | Choose the command from this list to send to the globe, then press the `sendCommand` button to send it as many times as you wish. |

Author: It does do that as well. The idea was select the command and it gets sent instantly when selected, then if you want to repeat the command the button sends it as many times as you wish. The result is less controls on a sitemap that allow access to all the commands, which most people will never use the commands.

> | command | String | Choose the command from this list to send to the globe, then press the `sendCommand` button to send it as many times as you wish. |

Reviewer: Understood. The problem I see here is that you are implementing an UI feature in the data model, which isn't really a good approach. Depending on the UI, the commands are rendered as individual buttons. I think you refer to a drop down field?

> | command | String | Choose the command from this list to send to the globe, then press the `sendCommand` button to send it as many times as you wish. |

Author: I use a sitemap with a selection item for testing, but I actually never use these myself. Having looked at the commands, the only ones which can not be done any other way is mode_speed_up and mode_speed_down. 

So I guess the choices are:
1. Throw away the two channels and remove the code.
2. Leave as is.

If the code is thrown away it can then be added by anyone in the future to move to using trigger? channels which is probably the way forward having not used them before.

> | command | String | Choose the command from this list to send to the globe, then press the `sendCommand` button to send it as many times as you wish. |

Reviewer: Did you try `Switch`?
`Switch item=Dusk_LED label="Sonne"` is rendered like this in Basic UI:
![grafik](https://user-images.githubusercontent.com/1894841/102191572-e2ceaf00-3eb9-11eb-9909-dde0f58cf62e.png)

Would that be practical for your use case?

> | command | String | Choose the command from this list to send to the globe, then press the `sendCommand` button to send it as many times as you wish. |

Author: Yes that would be practical as it is a 1 way protocol, just like a dumb IR remote. The switch should be a momentary action and not a latching ON and OFF state. What is the best way to implement that so it appears in the new UI by default? NOTE I changed the channels to Switch types, however I am thinking you may be suggesting another way?

> | command | String | Choose the command from this list to send to the globe, then press the `sendCommand` button to send it as many times as you wish. |

Reviewer: To achieve the button view as in above screenshot, you need to make the Channel of type `String` and define state options in the XML file. Each state option will be rendered to a button. In the `.sitemap` file, the type must be `Switch`: `Switch item=Dusk_LED label="Sonne"`

> | command | String | Choose the command from this list to send to the globe, then press the `sendCommand` button to send it as many times as you wish. |

Author: I made the change and it works as a drop down list in the new V3 UI, I tried to get the switches as per your post by using the UI to change the item to a Switch and all it gave was a ON and OFF ability. Anyway it works, I'm happy with it as I don't use the feature. Thanks for your help with this area and also doing the review.

> | command | String | Choose the command from this list to send to the globe, then press the `sendCommand` button to send it as many times as you wish. |

Reviewer: Ok, I'm not aware that there is any difference in Basic UI concerning the rendering of buttons between 2.5 and 3.x...

| sendCommand | Switch | Sends the selected command. |

## Note Regarding Transmission Delays

If you have lots of globes and openHAB turns them all on, you may notice a delay that causes the globes to turn on one by one and the delay can add up when a lot of globes are installed in your house.
This is caused by the time it takes to transmit the desired setting to the globe * by how many times the hub repeats transmitting the setting.
Since it takes around 2.8ms for a setting to be transmitted, if the firmware is set to repeat the packets 50 times it would then take 2.8*50 = 140ms before the next globe starts to have its new state transmitted by the hub.
You can reduce the packet repeats to speed up the response of this binding and the hub, but a few settings can be tweaked.

Settings for the radio tab found in the esp control panel using your browser, mine are:

Packet repeats = 12 (if you only turn 1 globe on or off it uses this value)
Packet repeat throttle threshold= 200
Packet repeat throttle sensitivity = 0

> Packet repeat minimum = 8 (When turning multiple globes on and off it will use this value as it throttles the repeats back to reduce latency/delay between each globe)

Author: DONE.

## Important for Textual Configuration

The binding requires things to have a specific format for the unique ID, the auto discovery does this for you.

```
espmilighthub:ThingType:broker:ThingUID
```

You need to add the Device ID and Group ID together to create the things unique ID.
The DeviceID is different for each remotes.
The GroupID can be 0 (all channels on the remote), or 1 to 8 for each of the individual channels on the remote).
If you do not understand this please use auto discovery to do it for you. 

The formula is
DeviceID + GroupID = ThingUID

For example:

| Device ID | Group ID |ThingUID  | 
|-----------|----------|----------|
| 0xC210    | 1        | 0xC2101  | 
| 0xB4CA    | 4        | 0xB4CA4  | 
| 0xB4CA    | 8        | 0xB4CA8  |

> | 0xB4CA    | 0        | 0xB4CA0  |

Author: DONE.

# EspMilightHub Binding

This binding allows an open source esp8266 based bridge to automatically find and add Milight globes.

> The hubs can be built from 2 ready made boards and  only need connecting with 7 wires. 

Reviewer 2: ```suggestion
The hubs can be built from 2 ready made boards and only need connecting with 7 wires. 
```

| `colourTemperature` | Dimmer | Change from cool to warm white with this control. |
| `colour` | Color | Allows you to change the colour, brightness and saturation of the globe. |
| `discoMode` | String | Switch to a Disco mode directly from a drop down list. |
| `bulbMode` | String (read only) | Displays the mode the bulb is currently in so that rules can determine if the globe is white, a color, disco modes or night mode are selected. |
| `command` | String | Sends the raw commands that the buttons on a remote send. |

## Note Regarding Transmission Delays

If you have lots of globes and openHAB turns them all on, you may notice a delay that causes the globes to turn on one by one and the delay can add up when a lot of globes are installed in your house.
This is caused by the time it takes to transmit the desired setting to the globe multiplied by how many times the hub repeats transmitting the setting.
Since it takes around 2.8ms for a setting to be transmitted, if the firmware is set to repeat the packets 50 times it would then take 2.8*50 = 140ms before the next globe starts to have its new state transmitted by the hub.

> You can reduce the packet repeats to speed up the response of this binding and the hub, but a few settings can be tweaked.

Reviewer 2: ```suggestion
You can reduce the packet repeats to speed up the response of this binding and the hub by tweaking a few settings.
```

> Settings for the radio tab found in the esp control panel using your browser, mine are:

Reviewer 2: ```suggestion
Settings can be found on the radio tab in the esp control panel using your browser.
Suggested settings are as follows:
```

+ Packet repeats = 12 (if you only turn 1 globe on or off it uses this value)
+ Packet repeat throttle threshold = 200
+ Packet repeat throttle sensitivity = 0
+ Packet repeat minimum = 8 (When turning multiple globes on and off it will use this value as it throttles the repeats back to reduce latency/delay between each globe)

## Important for Textual Configuration

This binding requires things to have a specific format for the unique ID, the auto discovery does this for you.

If doing textual configuration you need to add the Device ID and Group ID together to create the things unique ID.
The DeviceID is different for each remote.
The GroupID can be 0 (all channels on the remote), or 1 to 8 for each of the individual channels on the remote).
If you do not understand this please use auto discovery to do it for you. 

The formula is
DeviceID + GroupID = ThingUID

For example:

| Device ID | Group ID |ThingUID  | 
|-----------|----------|----------|
| 0xE6C     | 4        | 0xE6C4   | 
| 0xB4CA    | 4        | 0xB4CA4  | 
| 0xB4CA    | 8        | 0xB4CA8  |
| 0xB4CA    | 0        | 0xB4CA0  |

## Full Example

> To use these examples for textual configuration, you must have already have a configured a MQTT `broker` thing and know its unique ID.

Reviewer 2: ```suggestion
To use these examples for textual configuration, you must already have a configured a MQTT `broker` thing and know its unique ID.
```
